File: exam_at_home/1/annoying_queue.py
#
# author: jamie
# email: jinjiedeng.jjd@gmail.com
#
from collections import deque

def Solve (n, m):
    queue = deque()
    count_enqueue = [0] * n
    count_dequeue = [0] * n
    for _ in range(m):
        op = list(map(int, input().split()))
        if op[0] == 0: # enqueue
            queue.append(op[1])
            count_enqueue[op[1]] += 1
        elif op[0] == 1: # dequeue
            while len(queue) and not (count_enqueue[queue[0]] == 1 and count_dequeue[queue[0]] == 0):
                count_enqueue[queue[0]] -= 1
                count_dequeue[queue[0]] -= 1
                queue.popleft()
            if len(queue):
                count_enqueue[queue[0]] -= 1
                print(queue.popleft())
            else:
                print(-1)
        else: # leave
            count_dequeue[op[1]] += 1

if __name__ == "__main__":
    n, m = map(int, input().split())
    Solve(n, m)

File: exam_at_home/1/annoying_queue.cc
/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <deque>
#include <iostream>
#include <vector>

int main ()
{
  int64_t n, m;
  std::cin >> n >> m;
  std::vector<int64_t> count_enqueue(n), count_dequeue(n);
  std::deque<int64_t> queue;
  while (m--)
  {
    int64_t op, id;
    std::cin >> op;
    if (op == 0)
    {
      std::cin >> id;
      queue.push_back(id);
      ++count_enqueue[id];
    }
    else if (op == 1)
    {
      while (!queue.empty() && !(count_enqueue[queue.front()] == 1 && count_dequeue[queue.front()] == 0))
      {
        --count_enqueue[queue.front()];
        --count_dequeue[queue.front()];
        queue.pop_front();
      }
      if (!queue.empty())
      {
        std::cout << queue.front() << "\n";
        --count_enqueue[queue.front()];
        queue.pop_front();
      }
      else
      {
        std::cout << "-1\n";
      }
    }
    else
    {
      std::cin >> id;
      ++count_dequeue[id];
    }
  }
  return 0;
}